overlord/snapstate/snapmgr_test.go

```diff
@@ -930,28 +938,159 @@ func (s *snapmgrTestSuite) TestRemoveIntegration(c *C) {
 			name: "/snap/some-snap/7",
 		},
 		fakeOp{
+			op:    "remove-profiles:Doing",
+			name:  "some-snap",
+			revno: 7,
+		},
+		fakeOp{
 			op:   "remove-snap-data",
 			name: "/snap/some-snap/7",
 		},
 		fakeOp{
 			op:   "remove-snap-files",
 			name: "/snap/some-snap/7",
 		},
+		fakeOp{
+			op:   "discard-conns:Doing",
+			name: "some-snap",
+		},
 	}
 	c.Assert(s.fakeBackend.ops, DeepEquals, expected)
 
 	// verify snapSetup info
 	tasks := ts.Tasks()
-	// snap-setup is in discard-snap above discard-conns.
-	task := tasks[len(tasks)-2]
-	var ss snapstate.SnapSetup
-	err = task.Get("snap-setup", &ss)
-	c.Assert(err, IsNil)
-	c.Assert(ss, DeepEquals, snapstate.SnapSetup{
-		Name:     "some-snap",
-		Revision: 7,
+	for _, t := range tasks {
+		ss, err := snapstate.TaskSnapSetup(t)
+		c.Assert(err, IsNil)
+
+		var expSnapSetup *snapstate.SnapSetup
+		if t.Kind() == "discard-conns" {
+			expSnapSetup = &snapstate.SnapSetup{
+				Name: "some-snap",
+			}
+		} else {
+			expSnapSetup = &snapstate.SnapSetup{
+				Name:     "some-snap",
+				Revision: 7,
+			}
+		}
+		c.Check(ss, DeepEquals, expSnapSetup, Commentf(t.Kind()))
+	}
+
+	// verify snaps in the system state
+	var snapst snapstate.SnapState
+	err = snapstate.Get(s.state, "some-snap", &snapst)
+	c.Assert(err, Equals, state.ErrNoState)
+}
+
+func (s *snapmgrTestSuite) TestRemoveWithManyRevisionsIntegration(c *C) {
+	si3 := snap.SideInfo{
+		OfficialName: "some-snap",
+		Revision:     3,
+	}
+
+	si5 := snap.SideInfo{
+		OfficialName: "some-snap",
+		Revision:     5,
+	}
+
+	si7 := snap.SideInfo{
+		OfficialName: "some-snap",
+		Revision:     7,
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	snapstate.Set(s.state, "some-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{&si5, &si3, &si7},
 	})
 
+	chg := s.state.NewChange("remove", "remove a snap")
+	ts, err := snapstate.Remove(s.state, "some-snap", 0)
+	c.Assert(err, IsNil)
+	chg.AddAll(ts)
+
+	s.state.Unlock()
+	defer s.snapmgr.Stop()
+	s.settle()
+	s.state.Lock()
+
+	c.Assert(s.fakeBackend.ops, HasLen, 10)
+	expected := []fakeOp{
+		{
+			op:     "can-remove",
+			name:   "/snap/some-snap/7",
+			active: true,
+		},
+		{
+			op:   "unlink-snap",
+			name: "/snap/some-snap/7",
+		},
+		{
+			op:    "remove-profiles:Doing",
```

Comment:
unlink-snap calls snappy.UnlinkSnap which calls removePackageServices (which afaict is also responsible for first stopping them)